base_report_to_printer/models/printing_server.py

```diff
@@ -0,0 +1,232 @@
+        except:
+            message = _("Failed to connect to the CUPS server on %s:%s. "
+                        "Check that the CUPS server is running and that "
+                        "you can reach it from the Odoo server.") % (
+                            self.address, self.port)
+            _logger.warning(message)
+            if raise_on_error:
+                raise exceptions.UserError(message)
+
+        return connection
+
+    @api.multi
+    def action_update_printers(self):
+        return self.update_printers()
+
+    @api.multi
+    def update_printers(self, domain=None, raise_on_error=False):
+        mapping = {
+            3: 'available',
+            4: 'printing',
+            5: 'error'
+        }
+
+        if domain is None:
+            domain = []
+
+        servers = self
+        if not self:
+            servers = self.search(domain)
+
+        res = True
+        for server in servers:
+            connection = server._open_connection(raise_on_error=raise_on_error)
+            if not connection:
+                server.printer_ids.write({'status': 'server-error'})
+                res = False
+                continue
+
+            # Update Printers
+            printers = connection.getPrinters()
+            existing_printers = dict([
+                (printer.system_name, printer)
+                for printer in server.printer_ids
+            ])
+            updated_printers = []
+            for name, printer_info in printers.iteritems():
+                printer_values = {
+                    'name': printer_info['printer-info'],
+                    'system_name': name,
+                    'server_id': server.id,
+                    'uri': printer_info.get('device-uri', ''),
+                    'status': mapping.get(printer_info.get('printer-state'),
+                                          'unknown'),
+                    'status_message': printer_info.get(
+                        'printer-state-message', ''),
+                    'model': printer_info.get('printer-make-and-model', False),
+                    'location': printer_info.get('printer-location', False),
+                }
+                updated_printers.append(name)
+                if name not in existing_printers.keys():
+                    self.env['printing.printer'].sudo().create(printer_values)
+                else:
+                    existing_printers[name].sudo().write(printer_values)
+
+            # Set printers not found as unavailable
+            server.printer_ids.filtered(
+                lambda record: record.system_name not in updated_printers)\
+                .write({'status': 'unavailable'})
+
+        return res
+
+    @api.multi
+    def action_update_jobs(self):
+        if not self:
+            self = self.search([])
+        return self.update_jobs()
+
+    @api.multi
+    def update_jobs(self, which='all', first_job_id=-1):
+        job_obj = self.env['printing.job'].sudo()
```

Comment:
Why was this not possible without sudo? I find this incredibly dangerous to blanket sudo an object for such a long method.